staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+				ValsEqualThemselvesAndDataLiteral("val3", "val4", "false"),
+				"val1 != val4",
+				"type(val1) == bool",
+			},
+		},
+		{name: "duration format",
+			obj:    objs("1h2m3s4ms", "1h2m3s4ms"),
+			schema: schemas(durationFormat, durationFormat),
+			valid: []string{
+				ValsEqualThemselvesAndDataLiteral("val1", "val2", "duration('1h2m3s4ms')"),
+				"val1 == duration('1h2m') + duration('3s4ms')",
+				"val1.getHours() == 1",
+				"val1.getMinutes() == 62",
+				"val1.getSeconds() == 3723",
+				"val1.getMilliseconds() == 3723004",
+				"type(val1) == google.protobuf.Duration",
+			},
+		},
+		{name: "date format",
+			obj:    objs("1997-07-16", "1997-07-16"),
+			schema: schemas(dateFormat, dateFormat),
+			valid: []string{
+				ValsEqualThemselvesAndDataLiteral("val1", "val2", "timestamp('1997-07-16T00:00:00.000Z')"),
+				"val1.getDate() == 16",
+				"val1.getMonth() == 06", // zero based indexing
+				"val1.getFullYear() == 1997",
+				"type(val1) == google.protobuf.Timestamp",
+			},
+		},
+		{name: "date-time format",
+			obj:    objs("2011-08-18T19:03:37.010000000+01:00", "2011-08-18T19:03:37.010000000+01:00"),
+			schema: schemas(dateTimeFormat, dateTimeFormat),
+			valid: []string{
+				ValsEqualThemselvesAndDataLiteral("val1", "val2", "timestamp('2011-08-18T19:03:37.010+01:00')"),
+				"val1 == timestamp('2011-08-18T00:00:00.000+01:00') + duration('19h3m37s10ms')",
+				"val1.getDate('01:00') == 18",
+				"val1.getMonth('01:00') == 7", // zero based indexing
+				"val1.getFullYear('01:00') == 2011",
+				"val1.getHours('01:00') == 19",
+				"val1.getMinutes('01:00') == 03",
+				"val1.getSeconds('01:00') == 37",
+				"val1.getMilliseconds('01:00') == 10",
+				"val1.getHours('UTC') == 18", // TZ in string is 1hr off of UTC
+				"type(val1) == google.protobuf.Timestamp",
+			},
+		},
+		{name: "enums",
+			obj: map[string]interface{}{"enumStr": "Pending"},
+			schema: objectTypePtr(map[string]schema.Structural{"enumStr": {
+				Generic: schema.Generic{
+					Type: "string",
+				},
+				ValueValidation: &schema.ValueValidation{
+					Enum: []schema.JSON{
+						{Object: "Pending"},
+						{Object: "Available"},
+						{Object: "Bound"},
+						{Object: "Released"},
+						{Object: "Failed"},
+					},
+				},
+			}}),
+			valid: []string{
+				"enumStr == 'Pending'",
+				"enumStr in ['Pending', 'Available']",
+			},
+		},
+		{name: "conversions",
+			obj:    objs(int64(10), 10.0, 10.49, 10.5, true, "10", "MTA="),
+			schema: schemas(integerType, numberType, numberType, numberType, booleanType, stringType, byteType),
+			valid: []string{
+				"int(val2) == val1",
+				"int(val3) == val1",
+				"int(val4) == val1",
+				"int(val6) == val1",
+				"double(val1) == val2",
+				"double(val6) == val2",
+				"bytes(val6) == val7",
+				"string(val1) == val6",
+				"string(val2) == '10'",
+				"string(val3) == '10.49'",
+				"string(val4) == '10.5'",
+				"string(val5) == 'true'",
+				"string(val7) == val6",
```

Comment:
we could also test timestamp and duration conversions to string

Reply:
+1, Added